---
title: "Identification of Gene Signature of Breast Cancer Subtypes resistant to CDK4/6 inhibitors"
author: "Daehwan Kim and Prasanna Alluri"
date: "Feburary 11, 2018"
output: html_document
---

# Initialization

Delete R objects

```{r}
rm(list = ls())
```

Load R libraries

```{r message = FALSE}
library(affy)
library(affyPLM)
library(GEOquery)
library(genefilter)
library(cluster)
# library(WGCNA)
```

Define utility functions
```{r}
printf <- function(...) print(sprintf(...))
```

Load gene expression data (microarray) under GSE12790. The data is already included in the Git repository (https://github.com/infphilo/cowork_pa), so there is no need to download the data.
```{r message = FALSE, warning=FALSE}
geo_fname <- "GSE12790_series_matrix.txt.gz"
remote_fname <- paste("ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE12nnn/GSE12790/matrix/", geo_fname, sep="")
if(!file.exists(geo_fname)) download.file(remotefile, geo_fname)
hoeflich_2009 <- getGEO(filename=geo_fname)
```

Change the names of cell subtypes so that they match those of Finn et al. 2009
```{r}
pd <- pData(hoeflich_2009) # phenotype data
pd$cellLine <- gsub("MCF10a_Null_Vector_rep1", "MCF-10A", pd$title)
pd$cellLine <- gsub(".*_", "", pd$cellLine)
name_change <- c("MDA-MB-175VII", "MDA-MB-175", 
                 "ZR75-30", "ZR-75-30",
                 "MDA-MB-134VI", "MDA-MB-134",
                 "HCC202", "HCC-202",
                 "EFM19", "EFM-19",
                 "HCC1500", "HCC-1500",
                 "HCC1419", "HCC-1419",
                 "HCC38", "HCC-38",
                 "HCC2218", "HCC-2218",
                 "ZR75-1", "ZR-75-1",
                 "T47D", "T47-D",
                 "MCF7", "MCF-7",
                 "BT20", "BT-20",
                 "MDA-MB-435s", "MDA-MB-435",
                 "BT474", "BT-474",
                 "SKBR3", "SK-BR-3",
                 "KPL1", "KPL-1",
                 "HCC1143", "HCC-1143",
                 "HCC1395", "HCC-1395",
                 "Hs578T", "Hs578t",
                 "HCC1569", "HCC-1569",
                 "HCC1937", "HCC-1937",
                 "HCC1954", "HCC-1954",
                 "HCC70", "HCC-70")
for(i in 1:(length(name_change)/2)) {
  pd$cellLine <- gsub(name_change[i*2-1], name_change[i*2], pd$cellLine)
}
hoeflich_2009$cellLine <- pd$cellLine
```

Identify and use the cell lines from Hoeflich et al. 2009 that match Finn et al. 2009
```{r}
finn_2009 <- read.table("Finn_2009.txt", header = TRUE, sep = "\t")
selected_cols <- which(pd$cellLine %in% finn_2009$cellLine)
hoeflich_2009 <- hoeflich_2009[,selected_cols]
```

Transform gene expression (1 to `r as.integer(2^16 - 1)`) to log scale (0 to 16).
For example, gene expression values at `r ` are as follows.
```{r echo = FALSE}
exprs(hoeflich_2009)[1,]
```

```{r}
exprs(hoeflich_2009) <- log2(exprs(hoeflich_2009))
```
The log transformed gene expression values are:
`r exprs(hoeflich_2009)[1,]`

Add GEO ID to Finn et al. 2009 data.
```{r}
h2009_pd <- pData(hoeflich_2009)
# DK - there might be a better way to merge two data frames??
geo_subtype <- h2009_pd[, c("geo_accession", "cellLine")]
rownames(geo_subtype) <- NULL
finn_2009 <- merge(finn_2009, geo_subtype)
finn_2009 <- finn_2009[order(finn_2009$geo_accession),]
```

Identify genes that are differentially expressed either positively or negatively across `r nrow(finn_2009)` breast cancer subtypes.
```{r eval = FALSE}
h2009_exprs <- exprs(hoeflich_2009)
locus_names <- rownames(h2009_exprs)

stopifnot(finn_2009$geo_accession == colnames(h2009_exprs))
ic50s <- finn_2009$IC50
corr_table <- data.frame(matrix(nrow=nrow(h2009_exprs), ncol=2))
colnames(corr_table) <- c("locus", "corr")
for(i in 1:nrow(h2009_exprs)) {
  # Calculate correlation
  gene_exprs <- as.numeric(h2009_exprs[i,])
  corr_table[i, 1] = locus_names[i]
  corr_table[i, 2] = cor(ic50s, gene_exprs, method="pearson")
  if(i %% 1000 == 0) {
    printf("Calculated correlation values of %d loci", i)
  }
}

corr_table <- corr_table[order(corr_table$corr),]
head(corr_table)
tail(corr_table)
hist(corr_table[,2])
```

Additional codes:
```{r eval = FALSE}
# raw
# exprs(raw)[1:5,1:5]
# summary(exprs(raw)[,1:5])
# mva.pairs(exprs(raw)[,1:4], log.it = TRUE)

# dim(raw)
# raw <- varFilter(raw)
# dim(raw)

# features
features <- fData(raw)
stopifnot(all(rownames(features) == rownames(exprs(raw))))

# colnames(fData(raw))
# features[grep("TP53", features$`Gene Symbol`),]
# features[which(features$`Gene Symbol` == "TP53"),]
# features[match("TP53", features$`Gene Symbol`),]
# which(features$`Gene Symbol` == "TP53")

boxplot(exprs(raw))

E <- log2(exprs(raw))
par(mfrow=c(1,2))
plot(E[1274,], xlab="Array Index ",
     col="steelblue",pch=16,
     ylab="Normalised Expression",main=rownames(E)[1274])
plot(E[10723,], xlab="Array Index ",
     col="steelblue",pch=16,
     ylab="Normalised Expression",main=rownames(E)[10723])
cor(E[1274,],E[10723,], use = "complete.obs")

# Calcuate Euclidean distances among samples
euc.dist <- dist(t(exprs(raw)))

# corMat <-cor(exprs(raw))
# cor.dist <- as.dist(1 - corMat)

clust <- hclust(euc.dist)
clust
par(mfrow=c(1,1))

sample_group <- pd$characteristics_ch1
group_colors <- sample_group
levels(group_colors) <- c("yellow","blue", "red", "green", "gray", "black", "orange", "ivory")

clust.euclid <- plot(clust, labels=pd$title, cex=0.5)
# plotDendroAndColors(clust.euclid, colors=group_colors)

# Examples

targetsFile <- "estrogen/estrogen.txt"
pd <- read.AnnotatedDataFrame(targetsFile, header=TRUE, sep="", row.names=1)
raw <-ReadAffy(celfile.path="estrogen", filenames=rownames(pData(pd)), phenoData = pd)
boxplot(raw, col="red", las=2)

par(mfrow=c(2,1))
hist(log2(pm(raw[,1])), breaks=100, col="steelblue", main="PM", xlim=c(4,14))
hist(log2(mm(raw[,1])), breaks=100, col="steelblue", main="MM", xlim=c(4,14))

mva.pairs(pm(raw)[,1:4], plot.method="smoothScatter")

plmset <- fitPLM(raw)
NUSE(plmset,las=2)
RLE(plmset,las=2)

par(mfrow=c(2,4))
image(raw[,1])
image(raw[,2])
image(raw[,3])
image(raw[,4])
image(raw[,5])
image(raw[,6])
image(raw[,7])
image(raw[,8])

eset <- rma(raw)
head(exprs(eset))
summary(exprs(eset))
boxplot(exprs(eset),las=2)
mva.pairs(exprs(eset)[,1:4],log.it = FALSE,plot.method="smoothScatter")

```