---
title: "Identification of Gene Signature of Breast Cancer Subtypes Resistant to CDK4/6 Inhibitors"
author: "Daehwan Kim and Prasanna Alluri"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

# Outline

Steps involved in discovery of new drugs that reverse CDK4/6 inhibitor resistance:

(1) Develop CDK4/6 inhibitor resistance signature 

(2) Screen published databases to identify drugs that reverse the signature identified in step 1 (identical to Nature Communications paper we discussed) 

(3) Validate findings from step 2 in the lab 


# Initialization

Initialize R session by deleting existing R objects.

```{r}
rm(list = ls())
```

Load R libraries needed to anlayze microarray data and visualize results.

```{r message = FALSE}
library(affy)
library(affyPLM)
library(GEOquery)
library(genefilter)
library(cluster)
library(randomForest)
# library(WGCNA)
```

Define utility functions.
```{r}
printf <- function(...) print(sprintf(...))
```

Load gene expression data (microarray) under GSE12790, which is already included in the Git repository (https://github.com/infphilo/cowork_pa).
```{r message = FALSE, warning=FALSE}
geo_fname <- "GSE12790_series_matrix.txt.gz"
remote_fname <- paste("ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE12nnn/GSE12790/matrix/", geo_fname, sep="")
if(!file.exists(geo_fname)) download.file(remotefile, geo_fname)
hoeflich_2009 <- getGEO(filename=geo_fname)
```

Change the names of cell subtypes so that they match those of Finn et al. 2009
```{r}
pd <- pData(hoeflich_2009) # phenotype data
pd$cellLine <- gsub("MCF10a_Null_Vector_rep1", "MCF-10A", pd$title)
pd$cellLine <- gsub(".*_", "", pd$cellLine)
name_change <- c("MDA-MB-175VII", "MDA-MB-175", 
                 "ZR75-30", "ZR-75-30",
                 "MDA-MB-134VI", "MDA-MB-134",
                 "HCC202", "HCC-202",
                 "EFM19", "EFM-19",
                 "HCC1500", "HCC-1500",
                 "HCC1419", "HCC-1419",
                 "HCC38", "HCC-38",
                 "HCC2218", "HCC-2218",
                 "ZR75-1", "ZR-75-1",
                 "T47D", "T47-D",
                 "MCF7", "MCF-7",
                 "BT20", "BT-20",
                 "MDA-MB-435s", "MDA-MB-435",
                 "BT474", "BT-474",
                 "SKBR3", "SK-BR-3",
                 "KPL1", "KPL-1",
                 "HCC1143", "HCC-1143",
                 "HCC1395", "HCC-1395",
                 "Hs578T", "Hs578t",
                 "HCC1569", "HCC-1569",
                 "HCC1937", "HCC-1937",
                 "HCC1954", "HCC-1954",
                 "HCC70", "HCC-70")
for(i in 1:(length(name_change)/2)) {
  pd$cellLine <- gsub(name_change[i*2-1], name_change[i*2], pd$cellLine)
}
hoeflich_2009$cellLine <- pd$cellLine
```

Load a list of samples analyzed in Finn et al. 2009,
```{r}
finn_2009 <- read.table("Finn_2009.txt", header = TRUE, sep = "\t")
```
in which there are `r nrow(finn_2009)` subtypes. Unfortunately, their gene expression data is not publically available.

Identify and use the cell lines available from Hoeflich et al. 2009 that match Finn et al. 2009. 
```{r}
selected_cols <- which(pd$cellLine %in% finn_2009$cellLine)
hoeflich_2009 <- hoeflich_2009[,selected_cols]
```
There are `r ncol(hoeflich_2009)` subtypes common to both Hoeflich et al. 2009 and Finn et al. 2009, which we will use in our analysis of identifying gene signature resistant to CDK4/6 inhibitor (e.g. PD 0332991).

The following box plot shows the distribution of expression values across those `r ncol(hoeflich_2009)` subtypes.
```{r}
boxplot(log2(exprs(hoeflich_2009)), col="red", las=2)
```

The following MVA plot shows M (log ratio) as Y axis and A (mean average) as X axis of a pair of gene expression values (a, b) for each pair of four samples. A is defined as (log(a) + log(b)) / 2 and M is defined as log(a) - log(b).
```{r}
mva.pairs(exprs(hoeflich_2009)[,1:4], plot.method="smoothScatter")
```

```{r}
# Show images of gene expression values - it doesn't work now.
if(FALSE) {
ncol <- (ncol(hoeflich_2009) + 3) / 4
par(mfrow=c(4,ncol))
for(i in 1:ncol(hoeflich_2009)) {
  image(log2(exprs(hoeflich_2009[,i])))
}
}
```

Combine loci into groups that correspond to genes.
```{r}
h2009_exprs <- as.data.frame(exprs(hoeflich_2009))
locus2gene <- fData(hoeflich_2009)[c("ID", "Gene Symbol")]
stopifnot(rownames(h2009_exprs) == locus2gene$ID)
geneSymbol <- locus2gene$`Gene Symbol`
h2009_exprs <- cbind(geneSymbol, h2009_exprs)
```

Show gene expression values of four samples at WFDC2 and MAPK1.
```{r}
with(h2009_exprs, {
  h2009_exprs[geneSymbol == 'WFDC2' | geneSymbol == 'MAPK1', 1:5]
})
```

Merge gene expression values that belong to the same gene by the mean of the values.

```{r}
h2009_exprs <- aggregate(h2009_exprs[,2:ncol(h2009_exprs)], by = list(geneSymbol), FUN = mean)
rownames(h2009_exprs) <- h2009_exprs[,1]
h2009_exprs <- h2009_exprs[,-1]
```

```{r}
h2009_exprs[c("WFDC2", "MAPK1"), 1:4]
```

Transform gene expression values (1 to `r as.integer(2^16 - 1)`) to log scale (0 to 16).
```{r}
h2009_exprs <- log2(h2009_exprs)
```

For example, the log transformed gene expression values at **MAPK1** are as follows:
```{r echo = FALSE}
h2009_exprs["MAPK1",1:4]
```

Add GEO ID to Finn et al. 2009 data.
```{r}
h2009_pd <- pData(hoeflich_2009)
finn_2009 <- merge(finn_2009, h2009_pd[, c("geo_accession", "cellLine")])
```

Finn et al. 2009's data used in the analysis:
```{r echo = FALSE}
finn_2009[order(finn_2009$IC50),]
```

```{r echo = FALSE}
finn_2009 <- finn_2009[order(finn_2009$geo_accession),]
```

Identify genes that are differentially expressed either positively or negatively across `r nrow(finn_2009)` breast cancer subtypes.
```{r}
gene_names <- rownames(h2009_exprs)

stopifnot(finn_2009$geo_accession == colnames(h2009_exprs))
IC50 <- finn_2009$IC50
logIC50 <- log(IC50)
cor_table <- data.frame(matrix(nrow=nrow(h2009_exprs), ncol=2))
colnames(cor_table) <- c("locus", "cor")

cor_table[,1] <- gene_names
cor_table[,2] <- apply(h2009_exprs, 1, function(x) cor(x, logIC50, method="pearson"))
cor_order <- order(cor_table$cor)
```

Histogram of correlation values between IC50 and gene expression.
```{r}
hist(cor_table[,2])
```

```{r}
gene_names <- rownames(h2009_exprs)
most_pos_i <- cor_order[length(cor_order)]
plot(x=logIC50, y=h2009_exprs[most_pos_i,], xlab="IC50 (log)", ylab="Gene Expression (log)", main=paste("Most positively correlated at",gene_names[most_pos_i]))
abline(lm(t(h2009_exprs[most_pos_i,]) ~ logIC50))
most_neg_i <- cor_order[1]
plot(x=logIC50, y=h2009_exprs[most_neg_i,], xlab="IC50 (log)", ylab="Gene Expression (log)", main=paste("Most negatively correlated at",gene_names[most_neg_i]))
abline(lm(t(h2009_exprs[most_neg_i,]) ~ logIC50))
```

Show a heapmap.

```{r eval}
pos_th <- quantile(cor_table$cor, 0.999)
neg_th <- quantile(cor_table$cor, 0.001)

pos_rows <- which(cor_table$cor >= pos_th)
neg_rows <- which(cor_table$cor <= neg_th)
diff_rows <- c(pos_rows, neg_rows)

# Differentially expressed genes
diff_exprs <- as.matrix(h2009_exprs[diff_rows,])
colnames(diff_exprs) <- finn_2009$cellLine
heatmap(diff_exprs)

finn_2009[order(finn_2009$IC50), "cellLine"]
```

Create resistance signature (R-Sig) to mimic M-Sig developed in Zhao et al. 2015 using a random forest of 1,000 trees.
```{r}
samples <- t(diff_exprs)
stopifnot(rownames(samples) == finn_2009$cellLine)
samples <- cbind(samples, finn_2009$IC50)
colnames(samples)[ncol(samples)] <- "IC50"
colnames(samples) <- gsub(" ///.*", "", colnames(samples))
train <- sample(1:nrow(samples), nrow(samples) - 5)
rsig <- randomForest(IC50 ~ ., data = samples, subset = train, ntree = 1000)
print(rsig)
plot(rsig)
# print(importance(rsig, type = 2))
print("True IC50 values:")
samples[-train,"IC50"]
print("Predicted IC50 values:")
predict(rsig, samples[-train,])
```

#

# Supplementary Materials

A comprehensive tutorial for analysis of microarray data is available at http://bioinformatics-core-shared-training.github.io/microarray-analysis/
A tutorial for random forest is available at https://www.r-bloggers.com/random-forests-in-r/

Additional codes:
```{r eval = FALSE}
# raw <- varFilter(raw)

# features[grep("TP53", features$`Gene Symbol`),]
# features[which(features$`Gene Symbol` == "TP53"),]
# features[match("TP53", features$`Gene Symbol`),]
# which(features$`Gene Symbol` == "TP53")

# Calcuate Euclidean distances among samples
euc.dist <- dist(t(exprs(raw)))

# corMat <-cor(exprs(raw))
# cor.dist <- as.dist(1 - corMat)

clust <- hclust(euc.dist)
clust
par(mfrow=c(1,1))

sample_group <- pd$characteristics_ch1
group_colors <- sample_group
levels(group_colors) <- c("yellow","blue", "red", "green", "gray", "black", "orange", "ivory")

clust.euclid <- plot(clust, labels=pd$title, cex=0.5)
# plotDendroAndColors(clust.euclid, colors=group_colors)

# Examples

targetsFile <- "estrogen/estrogen.txt"
pd <- read.AnnotatedDataFrame(targetsFile, header=TRUE, sep="", row.names=1)
raw <-ReadAffy(celfile.path="estrogen", filenames=rownames(pData(pd)), phenoData = pd)
boxplot(raw, col="red", las=2)

par(mfrow=c(2,1))
hist(log2(pm(raw[,1])), breaks=100, col="steelblue", main="PM", xlim=c(4,14))
hist(log2(mm(raw[,1])), breaks=100, col="steelblue", main="MM", xlim=c(4,14))

mva.pairs(pm(raw)[,1:4], plot.method="smoothScatter")

plmset <- fitPLM(raw)
NUSE(plmset,las=2)
RLE(plmset,las=2)

par(mfrow=c(2,4))
image(raw[,1])
image(raw[,2])
image(raw[,3])
image(raw[,4])
image(raw[,5])
image(raw[,6])
image(raw[,7])
image(raw[,8])

eset <- rma(raw)
head(exprs(eset))
summary(exprs(eset))
boxplot(exprs(eset),las=2)
mva.pairs(exprs(eset)[,1:4],log.it = FALSE,plot.method="smoothScatter")

```