---
title: "Identification of Gene Signatures of Breast Cancer Subtypes Resistant to CDK4/6 Inhibitors"
author: "Daehwan Kim and Prasanna Alluri"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
bibliography: bibliography.bib
---

# Outline

Steps involved in discovery of new drugs that reverse CDK4/6 inhibitor resistance:

(1) Develop CDK4/6 inhibitor resistance signature.

(2) Screen published databases to identify drugs that reverse the signature identified in step 1 (identical to Chen et al. 2017).

(3) Validate findings from step 2 in the lab. 


# Initialization

Initialize R session by deleting existing R objects.

```{r}
rm(list = ls())
```

Load R libraries needed to analyze microarray data and visualize results.

```{r message = FALSE}
library(affy)
library(affyPLM)
library(GEOquery)
library(genefilter)
library(cluster)
library(randomForest)
library(plyr)
# library(WGCNA)
```

Define utility functions.
```{r}
printf <- function(...) print(sprintf(...))
```

Settings:
```{r}
fast_run <- FALSE
```

Load gene expression data (microarray) under GSE12790, which is already included in our GitHub repository (https://github.com/infphilo/cowork_pa).
```{r message = FALSE, warning=FALSE}
geo_fname <- "GSE12790_series_matrix.txt.gz"
remote_fname <- paste("ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE12nnn/GSE12790/matrix/", geo_fname, sep="")
if(!file.exists(geo_fname)) download.file(remotefile, geo_fname)
hoeflich_2009 <- getGEO(filename=geo_fname)
```

Change the names of cell subtypes so that they match those of @finn2009
```{r}
pd <- pData(hoeflich_2009) # phenotype data
pd$cellLine <- gsub("MCF10a_Null_Vector_rep1", "MCF-10A", pd$title)
pd$cellLine <- gsub(".*_", "", pd$cellLine)
name_change <- c("MDA-MB-175VII", "MDA-MB-175", 
                 "ZR75-30", "ZR-75-30",
                 "MDA-MB-134VI", "MDA-MB-134",
                 "HCC202", "HCC-202",
                 "EFM19", "EFM-19",
                 "HCC1500", "HCC-1500",
                 "HCC1419", "HCC-1419",
                 "HCC38", "HCC-38",
                 "HCC2218", "HCC-2218",
                 "ZR75-1", "ZR-75-1",
                 "T47D", "T47-D",
                 "MCF7", "MCF-7",
                 "BT20", "BT-20",
                 "MDA-MB-435s", "MDA-MB-435",
                 "BT474", "BT-474",
                 "SKBR3", "SK-BR-3",
                 "KPL1", "KPL-1",
                 "HCC1143", "HCC-1143",
                 "HCC1395", "HCC-1395",
                 "Hs578T", "Hs578t",
                 "HCC1569", "HCC-1569",
                 "HCC1937", "HCC-1937",
                 "HCC1954", "HCC-1954",
                 "HCC70", "HCC-70")
for(i in 1:(length(name_change)/2)) {
  pd$cellLine <- gsub(name_change[i*2-1], name_change[i*2], pd$cellLine)
}
hoeflich_2009$cellLine <- pd$cellLine
```

Load the list of samples analyzed in @finn2009.
```{r}
finn_2009 <- read.table("Finn_2009.txt", header = TRUE, sep = "\t")
```
The list includes `r nrow(finn_2009)` subtypes. Unfortunately, their gene expression data is not publically available.

Identify and use the cell lines available from @hoeflich2009 that match @finn2009. 
```{r}
selected_cols <- which(pd$cellLine %in% finn_2009$cellLine)
hoeflich_2009 <- hoeflich_2009[,selected_cols]
```
There are `r ncol(hoeflich_2009)` subtypes common to both @hoeflich2009 and @finn2009, which we will use in our analysis to identify gene signatures resistant to CDK4/6 inhibitor (e.g. PD 0332991).

The following box plot shows the distribution of gene expression values across those `r ncol(hoeflich_2009)` subtypes.
```{r}
boxplot(log2(exprs(hoeflich_2009)), col="red", las=2, names=hoeflich_2009$cellLine, cex.axis=0.8)
```

The following MVA plot shows M (log ratio) on Y axis and A (mean average) on X axis for a pair of gene expression values (a, b) for each pair of four samples. A is defined as (log(a) + log(b)) / 2 and M is defined as log(a) - log(b).
```{r}
tmp_exprs <- exprs(hoeflich_2009)
colnames(tmp_exprs) <- hoeflich_2009$cellLine
mva.pairs(tmp_exprs[, 1:4], plot.method="smoothScatter", cex=1.2)
```

```{r}
# Show images of gene expression values - it doesn't work now.
if(FALSE) {
ncol <- (ncol(hoeflich_2009) + 3) / 4
par(mfrow=c(4,ncol))
for(i in 1:ncol(hoeflich_2009)) {
  image(log2(exprs(hoeflich_2009[,i])))
}
}
```

Combine loci into groups that correspond to genes.
```{r}
h2009_exprs <- as.data.frame(exprs(hoeflich_2009))
locus2gene <- fData(hoeflich_2009)[c("ID", "Gene Symbol")]
stopifnot(rownames(h2009_exprs) == locus2gene$ID)
geneSymbol <- locus2gene$`Gene Symbol`
geneSymbol <- gsub(" ///.*", "", geneSymbol)
h2009_exprs <- cbind(geneSymbol, h2009_exprs)
```

Before gene expression values are combined, the following table shows gene expression values of four samples at WFDC2 and MAPK1, for example.
```{r}
with(h2009_exprs, {
  h2009_exprs[geneSymbol == 'WFDC2' | geneSymbol == 'MAPK1', 1:5]
})
```

Merge gene expression values that belong to the same gene by the mean of the values.

```{r}
h2009_exprs <- aggregate(h2009_exprs[,2:ncol(h2009_exprs)], by = list(geneSymbol), FUN = mean)
rownames(h2009_exprs) <- h2009_exprs[,1]
h2009_exprs <- h2009_exprs[,-1]
```

```{r}
h2009_exprs[c("WFDC2", "MAPK1"), 1:4]
```

Transform the gene expression values (`r min(h2009_exprs)` to `r format(max(h2009_exprs), digits=5)`) to a log scale (`r log2(min(h2009_exprs))` to `r log2(max(h2009_exprs))`).
```{r}
h2009_exprs <- log2(h2009_exprs)
```

For example, the log transformed gene expression values at MAPK1 are as follows:
```{r echo = FALSE}
h2009_exprs["MAPK1",1:4]
```

Add GEO ID to @finn2009 data.
```{r}
h2009_pd <- pData(hoeflich_2009)
finn_2009 <- merge(finn_2009, h2009_pd[, c("geo_accession", "cellLine")])
```

The list of samples analyzed in @finn2009 are:
```{r echo = FALSE}
finn_2009[order(finn_2009$IC50),]
```

```{r echo = FALSE}
finn_2009 <- finn_2009[order(finn_2009$geo_accession),]
```

Identify genes that are differentially expressed either positively or negatively across `r nrow(finn_2009)` breast cancer subtypes according to IC50 values.
```{r}
gene_names <- rownames(h2009_exprs)

stopifnot(finn_2009$geo_accession == colnames(h2009_exprs))
IC50 <- finn_2009$IC50
logIC50 <- log(IC50)
cor_table <- data.frame(matrix(nrow=length(gene_names), ncol=2))
colnames(cor_table) <- c("locus", "cor")

cor_table[,1] <- gene_names
cor_table[,2] <- apply(h2009_exprs, 1, function(x) cor(x, logIC50, method="pearson"))
cor_order <- order(cor_table$cor)
```

The following is a histogram of correlation values between IC50 and gene expression.
```{r}
hist(cor_table[,2])
```

```{r}
gene_names <- rownames(h2009_exprs)
par(mfrow = c(3,2))
for(i in cor_order[length(cor_order):(length(cor_order)-5)]) {
  plot(x=logIC50, y=h2009_exprs[i,], cex=0.5, xlab="IC50 (log)", ylab="Gene Expression (log)", main=paste("Most positively correlated at", gene_names[i]))
  abline(lm(t(h2009_exprs[i,]) ~ logIC50))
}
par(mfrow = c(3,2))
for(i in cor_order[1:6]) {
  plot(x=logIC50, y=h2009_exprs[i,], cex=0.5, xlab="IC50 (log)", ylab="Gene Expression (log)", main=paste("Most negatively correlated at", gene_names[i]))
  abline(lm(t(h2009_exprs[i,]) ~ logIC50))
}
```

The following shows a heatmap based on the top 0.1% most positively and negatively correlated genes.

```{r eval}
pos_th <- quantile(cor_table$cor, 0.999)
neg_th <- quantile(cor_table$cor, 0.001)

pos_rows <- which(cor_table$cor >= pos_th)
neg_rows <- which(cor_table$cor <= neg_th)
diff_rows <- c(pos_rows, neg_rows)

# Differentially expressed genes
diff_exprs <- as.matrix(h2009_exprs[diff_rows,])
colnames(diff_exprs) <- finn_2009$cellLine
heatmap(diff_exprs)

finn_2009[order(finn_2009$IC50), "cellLine"]
```

Create Resistance Signature (R-Sig) to mimic Metastasis Signature (M-Sig) developed in @zhao2015 using a random forest of 1,000 trees.
```{r}
samples <- t(diff_exprs)
stopifnot(rownames(samples) == finn_2009$cellLine)
samples <- cbind(samples, finn_2009$IC50)
colnames(samples)[ncol(samples)] <- "IC50"
train <- sample(1:nrow(samples), nrow(samples) - 5)
rsig <- randomForest(IC50 ~ ., data = samples, subset = train, ntree = 1000)
print(rsig)
plot(rsig)
# print(importance(rsig, type = 2))
```

True IC50 values:
```{r}
samples[-train, "IC50"]
```

Predicted IC50 values:
```{r}
predict(rsig, samples[-train,])
```

Load Library of Integrated Network-based Cellular Signatures (LINCS) database downloaded from 
http://lincs.hms.harvard.edu/. R data files such as *lincs_signatures_cmpd_landmark.RData* and *lincs_sig_info.csv* are downloaded from @chen2017. The following code is taken from @chen2017 (https://github.com/Bin-Chen-Lab/RGES) and modified.
```{r}
load("lincs_signatures_cmpd_landmark.RData")
lincs_sig_info <- read.csv("lincs_sig_info.csv")
lincs_sig_info <- subset(lincs_sig_info, id %in% colnames(lincs_signatures))
lincs_sig_info <- lincs_sig_info[!duplicated(lincs_sig_info$id),]

lincs_gene_info <- read.csv("lincs_landmark.csv")
stopifnot(rownames(lincs_signatures) == lincs_gene_info$gene_id)
rownames(lincs_signatures) <- lincs_gene_info$gene_symbol
```

LINCS contains expression changes of `r nrow(lincs_signatures)` genes across `r ncol(lincs_signatures)` perturbations in which there are `r length(unique(lincs_sig_info$pert_iname))` unique compounds, so the dimension of matrix is `r nrow(lincs_signatures)` x `r ncol(lincs_signatures)`. For example, the following table shows expression changes of 10 genes across 5 perturbations.

```{r}
lincs_signatures[1:10,1:5]
```

The following table shows related information:
```{r}
lincs_sig_info[lincs_sig_info$id %in% colnames(lincs_signatures)[1:5],]
```

```{r}
pos_th <- quantile(cor_table$cor, 0.99)
neg_th <- quantile(cor_table$cor, 0.01)
pos_rows <- which(cor_table$cor >= pos_th)
neg_rows <- which(cor_table$cor <= neg_th)

lincs_signatures_pos <- lincs_signatures[rownames(lincs_signatures) %in% gene_names[pos_rows],]
lincs_signatures_neg <- lincs_signatures[rownames(lincs_signatures) %in% gene_names[neg_rows],]
```
The number of genes that match positively and negatively correlated genes to IC50 values are `r nrow(lincs_signatures_pos)` and `r nrow(lincs_signatures_neg)`, respectively.

Calculate Reverse Gene Expression Score (RGES) developed by Chen et al. 2017.
```{r}
gene_list <- rownames(lincs_signatures)

cmap_score <- function(sig_up, sig_down, drug_signature) {
  num_genes <- nrow(drug_signature)

  if(length(sig_up) > 0) {
    up_tags_rank <- merge(drug_signature, sig_up, by.x = "id", by.y = 1)
    up_tags_position <- sort(up_tags_rank$rank)
  } else {
    up_tags_position <- NULL
  }
  
  if(length(sig_down) > 0) {
    down_tags_rank <- merge(drug_signature, sig_down, by.x = "id", by.y = 1)
    down_tags_position <- sort(down_tags_rank$rank)
  } else {
    down_tags_position <- NULL
  }
  
  num_tags_up <- length(up_tags_position)
  num_tags_down <- length(down_tags_position)
  
  ks_up <- 0
  if(num_tags_up > 1) {
    a_up <- max(sapply(1:num_tags_up, function(j) j / num_tags_up - up_tags_position[j] / num_genes))
    b_up <- max(sapply(1:num_tags_up, function(j) up_tags_position[j] / num_genes - (j-1) / num_tags_up))
    if(a_up > b_up) {
      ks_up <- a_up
    } else {
      ks_up <- -b_up
    }
  }
  
  ks_down <- 0
  if(num_tags_down > 1) {
    a_down <- max(sapply(1:num_tags_down, function(j) j / num_tags_down - down_tags_position [j] / num_genes))
    b_down <- max(sapply(1:num_tags_down, function(j) down_tags_position[j] / num_genes - (j-1) / num_tags_down))
    if(a_down > b_down) {
      ks_down <- a_down
    } else {
      ks_down <- -b_down
    }
  }
  
  connectivity_score <- ks_up - ks_down
  return(connectivity_score)
}
```

```{r eval=FALSE}
# Test cmap_score function
exp_signature <- data.frame(gene_list,  rank(-1 * lincs_signatures[,1], ties.method="random"))
colnames(exp_signature) <- c("id", "rank")
sorted_exp_signature <- exp_signature[order(exp_signature$rank),]
  
pos_signature <- sorted_exp_signature[1:50, "id"]
neg_signature <- sorted_exp_signature[(nrow(exp_signature)-49):nrow(exp_signature), "id"]
  
printf("Testing positive cmap score:")
print(cmap_score(pos_signature, c(), exp_signature))
  
printf("Testing negitive cmap score:")
print(cmap_score(c(), pos_signature, exp_signature))
  
printf("Testing positive cmap score2:")
print(cmap_score(c(), neg_signature, exp_signature))
  
printf("Testing negitive cmap score2:")
print(cmap_score(neg_signature, c(), exp_signature))
  
printf("Testing positive and negitive cmap score:")
print(cmap_score(pos_signature, neg_signature, exp_signature))
  
printf("Testing negitive and positive cmap score:")
print(cmap_score(neg_signature, pos_signature, exp_signature))
```
  
```{r}
if(fast_run) {
  exp_ids <- lincs_sig_info$id[1:1000]
} else {
  exp_ids <- lincs_sig_info$id
}

dz_genes_up <- gene_names[pos_rows]
dz_genes_down <- gene_names[neg_rows]

cmap_scores <- data.frame(matrix(nrow=length(exp_ids), ncol=2))
colnames(cmap_scores) <- c("exp_id", "score")
cmap_scores$exp_id <- exp_ids
for (i in 1:nrow(cmap_scores)) {
  exp_id <- cmap_scores[i,"exp_id"]
  cmap_exp_signature <- data.frame(gene_list,  rank(-1 * lincs_signatures[, as.character(exp_id)], ties.method="random"))
  colnames(cmap_exp_signature) <- c("id", "rank")
  cmap_scores[i, "score"] <- cmap_score(dz_genes_up, dz_genes_down, cmap_exp_signature)
}

cmap_score_info <- merge(cmap_scores, lincs_sig_info, by.x = "exp_id", by.y = "id")
cmap_score_info <- cmap_score_info[order(cmap_score_info$score),]
```

Top 50 perturbagen candidates:
```{r}
cmap_score_info[1:20,]
```

CMAP scores related to Palbociclib:
```{r}
cmap_score_info[cmap_score_info$exp_id %in% lincs_sig_info[lincs_sig_info$pert_iname == "palbociclib", "id"],]
```

```{r}
N_PERMUTATIONS <- ifelse(fast_run, 1000, 10000)
random_sig_ids <- sample(1:ncol(lincs_signatures), N_PERMUTATIONS, replace=T)
random_cmap_scores <- NULL
for(expr_id in random_sig_ids) {
  cmap_exp_signature <- data.frame(gene_list,  rank(-1 * lincs_signatures[, as.character(exp_id)], ties.method="random"))
  colnames(cmap_exp_signature) <- c("id", "rank")
  random_input_signature_genes <- sample(gene_names, (length(dz_genes_up) + length(dz_genes_down)))
  rand_dz_gene_up <- random_input_signature_genes[1:length(dz_genes_up)]
  rand_dz_gene_down <- random_input_signature_genes[(length(dz_genes_up)+1):length(random_input_signature_genes)]
  random_cmap_scores <- c(random_cmap_scores, cmap_score(rand_dz_gene_up, rand_dz_gene_down, cmap_exp_signature))
}

p <- sapply(cmap_scores$score, function(score) sum(random_cmap_scores < score) / length(random_cmap_scores))

padj <- p.adjust(p, "fdr")
results <- data.frame(id = exp_ids, cmap_score = cmap_scores$score, p, padj)
results <- merge(results, lincs_sig_info, by = "id")
results <- results[order(results$cmap_score),]
results[1:20,]
```

```{r}
# Code is from Chen et al. 2017 and modified
getsRGES <- function(RGES, cor, pert_dose, pert_time, diff, max_cor) {
  sRGES <- RGES
  pert_time <- ifelse(pert_time < 24, "short", "long")
  pert_dose <- ifelse(pert_dose < 10, "low", "high")
  if(pert_time == "short" & pert_dose == "low") {
    sRGES <- sRGES + diff[4]
  } else if(pert_dose ==  "low" & pert_time == "long") {
    sRGES <- sRGES + diff[2]
  } else if(pert_dose ==  "high" & pert_time == "short") {
    sRGES <- sRGES + diff[1]
  }
  return(sRGES * cor/max_cor)
}

lincs_drug_prediction <- results

# should use pert_dose > 0.01
lincs_drug_prediction_subset <- subset(lincs_drug_prediction,  pert_dose > 0 & pert_time %in% c(6, 24))

# pairs that share the same drug and cell id
lincs_drug_prediction_pairs <- merge(lincs_drug_prediction_subset, lincs_drug_prediction_subset, by=c("pert_iname", "cell_id")) 

# x is the reference
lincs_drug_prediction_pairs <- subset(lincs_drug_prediction_pairs, id.x != id.y & pert_time.x == 24 & pert_dose.x == 10) #, 
# difference of RGES to the reference 
lincs_drug_prediction_pairs$cmap_diff <- lincs_drug_prediction_pairs$cmap_score.x - lincs_drug_prediction_pairs$cmap_score.y
lincs_drug_prediction_pairs$dose <- round(log(lincs_drug_prediction_pairs$pert_dose.y, 2), 1)

# estimate difference
lincs_drug_prediction_pairs$dose_bin <- ifelse(lincs_drug_prediction_pairs$pert_dose.y < 10, "low", "high")
diff <- tapply(lincs_drug_prediction_pairs$cmap_diff, paste(lincs_drug_prediction_pairs$dose_bin, lincs_drug_prediction_pairs$pert_time.y), mean)

# ignore weighting cell lines
lincs_cell_line_weight <- read.csv("lincs_cell_line_weight.csv")
pred <- merge(lincs_drug_prediction, lincs_cell_line_weight, by.x="cell_id", by.y="lincs_cell_id")
pred$RGES <- sapply(1:nrow(pred), function(id) getsRGES(pred$cmap_score[id], pred$cor[id], pred$pert_dose[id], pred$pert_time[id], diff, max(pred$cor)))

cmpd_freq <- table(pred$pert_iname)
pred <- subset(pred, pert_iname %in% names(cmpd_freq[cmpd_freq > 0]))

pred_merged <- ddply(pred,  .(pert_iname),  summarise,
                     mean = mean(RGES),
                     n = length(RGES),
                     median = median(RGES),
                     sd = sd(RGES))
pred_merged$sRGES <- pred_merged$mean
pred_merged <- pred_merged[order(pred_merged$sRGES), ]
pred_merged[1:20,]
```

# Supplementary Materials

A comprehensive tutorial for analysis of microarray data is available at http://bioinformatics-core-shared-training.github.io/microarray-analysis/

A tutorial for using random forest is available at https://www.r-bloggers.com/random-forests-in-r/

Additional codes:
```{r eval = FALSE}
# raw <- varFilter(raw)

# features[grep("TP53", features$`Gene Symbol`),]
# features[which(features$`Gene Symbol` == "TP53"),]
# features[match("TP53", features$`Gene Symbol`),]
# which(features$`Gene Symbol` == "TP53")

# Calcuate Euclidean distances among samples
euc.dist <- dist(t(exprs(raw)))

# corMat <-cor(exprs(raw))
# cor.dist <- as.dist(1 - corMat)

clust <- hclust(euc.dist)
clust
par(mfrow=c(1,1))

sample_group <- pd$characteristics_ch1
group_colors <- sample_group
levels(group_colors) <- c("yellow","blue", "red", "green", "gray", "black", "orange", "ivory")

clust.euclid <- plot(clust, labels=pd$title, cex=0.5)
# plotDendroAndColors(clust.euclid, colors=group_colors)

# Examples

targetsFile <- "estrogen/estrogen.txt"
pd <- read.AnnotatedDataFrame(targetsFile, header=TRUE, sep="", row.names=1)
raw <-ReadAffy(celfile.path="estrogen", filenames=rownames(pData(pd)), phenoData = pd)
boxplot(raw, col="red", las=2)

par(mfrow=c(2,1))
hist(log2(pm(raw[,1])), breaks=100, col="steelblue", main="PM", xlim=c(4,14))
hist(log2(mm(raw[,1])), breaks=100, col="steelblue", main="MM", xlim=c(4,14))

mva.pairs(pm(raw)[,1:4], plot.method="smoothScatter")

plmset <- fitPLM(raw)
NUSE(plmset,las=2)
RLE(plmset,las=2)

par(mfrow=c(2,4))
image(raw[,1])
image(raw[,2])
image(raw[,3])
image(raw[,4])
image(raw[,5])
image(raw[,6])
image(raw[,7])
image(raw[,8])

eset <- rma(raw)
head(exprs(eset))
summary(exprs(eset))
boxplot(exprs(eset),las=2)
mva.pairs(exprs(eset)[,1:4],log.it = FALSE,plot.method="smoothScatter")

```
